src/rdkafka.c

```diff
@@ -2844,6 +2844,410 @@ rd_kafka_t *rd_kafka_new(rd_kafka_type_t type,
+                             rko_sf->rko_u.share_fetch.should_fetch
+                                 ? "should"
+                                 : "should not");
+                rd_kafka_q_enq(rkb->rkb_ops, rko_sf);
+        }
+        rd_kafka_rdunlock(rk);
+
+        RD_IF_FREE(selected_rkb, rd_kafka_broker_destroy);
+
+        return RD_KAFKA_OP_RES_HANDLED;
+}
+
+rd_kafka_error_t *rd_kafka_share_consume_batch(
+    rd_kafka_t *rk,
+    int timeout_ms,
+    /* There is some benefit to making this ***rkmessages and allocating it
+       within this function, but on the flipside this means that it will always
+       be allocated on the heap. */
+    rd_kafka_message_t **rkmessages /* out */,
+    size_t *rkmessages_size /* out */) {
+        rd_kafka_cgrp_t *rkcg;
+        rd_ts_t now             = rd_clock();
+        rd_ts_t abs_timeout     = rd_timeout_init0(now, timeout_ms);
+        size_t max_poll_records = (size_t)rk->rk_conf.share.max_poll_records;
+
+        if (!RD_KAFKA_IS_SHARE_CONSUMER(rk))
+                return rd_kafka_error_new(RD_KAFKA_RESP_ERR__INVALID_ARG,
+                                          "rd_kafka_share_consume_batch(): "
+                                          "rk is not a shared consumer");
+
+        if (unlikely(!(rkcg = rd_kafka_cgrp_get(rk))))
+                return rd_kafka_error_new(RD_KAFKA_RESP_ERR__STATE,
+                                          "rd_kafka_share_consume_batch(): "
+                                          "Consumer group not initialized");
+
+        /* If we have any pending items on the consumer queue, don't issue new
+         * requests, rather, deal with them first. 
+         * 
+         * TODO KIP-932:
+         * Above statement might be incorrect as we have to send all the pending
+         * acknowledgements irrespective of whether there are messages to be
+         * consumed or not.
+         */
+        if (likely(rd_kafka_q_len(rkcg->rkcg_q) == 0)) {
+                rd_kafka_dbg(rk, CGRP, "SHARE",
+                             "Issuing share fetch fanout to main thread with "
+                             "abs_timeout = %" PRId64,
+                             abs_timeout);
+                rd_kafka_share_fetch_fanout_with_backoff(rk, abs_timeout,
+                                                         0 /* no backoff */);
+        }
+
+        /* At this point, there's no reason to deviate from what we already do
+         * for returning multiple messages to the user, as the orchestration
+         * is handled by the main thread. Later on, we needed, we might need
+         * a custom loop if we need any changes. */
+        *rkmessages_size = rd_kafka_q_serve_share_rkmessages(
+            rkcg->rkcg_q, timeout_ms, /* Use this timeout directly as prior
+                             operations aren't blocking, so no need to
+                             re-convert the abs_timeout into a relative one.*/
+            rkmessages, max_poll_records);
```

Comment:
TODO: Change this to send all the messages present in the queue at once instead of sending only max_poll_records. How many messages to be sent to the user is driven by broker with AcquiredRecords field.